[TECU]: Add ISOBUS Heartbeat Message

ad3154 opened the pull request:
## Describe your changes

<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Added an interface which manages sending and receiving the ISOBUS heartbeat.

The heartbeat message is used to determine the integrity of the communication of messages and parameters being transmitted by a control function. There may be multiple instances of the heartbeat message on the network, and CFs are required transmit the message on request. As long as the heartbeat message is transmitted at the regular time interval and the sequence number increases through the valid range, then the heartbeat message indicates that the data source CF is operational and provides correct data in all its messages.

This interface is enabled by default, but can be disabled if you want to stop your heartbeat(s) or don't care about the safety-critical path of the machine.

https://www.isobus.net/isobus/pGNAndSPN/2507?type=PGN

## How has this been tested?

- [X] Added unit test for heartbeat interface
- [x] Tested with seeder example

GwnDaan commented on isobus/src/isobus_heartbeat.cpp:
```diff
@@ -0,0 +1,287 @@
+#include "isobus/isobus/isobus_heartbeat.hpp"
+
+#include "isobus/isobus/can_general_parameter_group_numbers.hpp"
+#include "isobus/isobus/can_network_manager.hpp"
```
thought: I think we have a circular dependency currently, correct me if I'm wrong. Maintaining (and testing) this interface should be easier if we can remove it. I'll leave a few more comments on how I would approach it

ad3154 replied:
Circular dependency now removed

GwnDaan commented on isobus/src/isobus_heartbeat.cpp:
```diff
@@ -0,0 +1,287 @@
+		return initialized;
+	}
+
+	void HeartbeatInterface::set_enabled(bool enable)
+	{
+		if ((!enable) && (enable != enabled))
+		{
+			LOG_DEBUG("[HB]: Disabling ISOBUS heartbeat interface.");
+			disable();
+		}
+		enabled = enable;
+	}
```
resolving circular-dependency: remove call to `disable()` here, and instead use an early return of some kind in the `update()` and `process_request_for_heartbeat()`. Although not ideal, it simplifies the states disabling/enabling the interface and helps us remove the circular dependency
```c++
if (!enabled) {
    return;
}
```

ad3154 replied:
I prefer adding to the top of the enclosing `if` checks personally, which accomplishes the same thing basically. Added where needed.

GwnDaan commented on isobus/src/isobus_heartbeat.cpp:
```diff
@@ -0,0 +1,287 @@
+					pgnRequestProtocol->register_request_for_repetition_rate_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_request_for_heartbeat, this);
+				}
+			}
+			CANNetworkManager::CANNetwork.add_any_control_function_parameter_group_number_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_rx_message, this);
```
resolving circular-dependency: Instead of adding a "any" callback for this PGN, we can call `process_rx_message()` directly from the CANNetworkManager

GwnDaan commented on isobus/src/isobus_heartbeat.cpp:
```diff
@@ -0,0 +1,287 @@
+
+namespace isobus
+{
+	void HeartbeatInterface::initialize()
+	{
+		if (enabled && !initialized)
+		{
+			auto internalControlFunctions = CANNetworkManager::CANNetwork.get_internal_control_functions();
+
+			for (auto &cf : internalControlFunctions)
+			{
+				auto pgnRequestProtocol = cf->get_pgn_request_protocol().lock();
+
+				if (nullptr != pgnRequestProtocol)
+				{
+					pgnRequestProtocol->register_request_for_repetition_rate_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_request_for_heartbeat, this);
+				}
+			}
+			CANNetworkManager::CANNetwork.add_any_control_function_parameter_group_number_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_rx_message, this);
+			initialized = true;
+		}
+	}
```
thought: theoretically there can't be any internal control function that `on_new_internal_control_function` doesn't catch? If so then this initialize function can be removed

ad3154 replied:
I think so? Though this reminds me that I need to implement the reverse, so when an internal control function is deleted I should clean up these registrations probably.

GwnDaan commented on isobus/src/isobus_heartbeat.cpp:
```diff
@@ -0,0 +1,287 @@
+		bool retVal = false;
+		const std::array<std::uint8_t, 1> buffer = { sequenceCounter };
+
+		retVal = CANNetworkManager::CANNetwork.send_can_message(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage),
```
resolving circular-dependency: this is the interesting part, to make this interface completely free from the network manager then we probably need a callback for sending messages like in the recently refactored transport protocols? I still don't really like that approach but I haven't found anything nicer yet

ad3154 replied:
Yeah, I dunno if I super love how we're doing the callback thing with those, but I agree I don't have an alterative to propose, so for the time being we might as well keep it consistent and do something similar...

GwnDaan replied:
Seems you forgot to implement `sendCANFrameCallback` here?

ad3154 replied:
Quite right, fixed.

GwnDaan commented on isobus/src/isobus_heartbeat.cpp:
```diff
@@ -0,0 +1,287 @@
+					LOG_ERROR("[HB]: Duplicate sequence counter received in heartbeat.");
+					interface->heartbeatErrorEventDispatcher.call(HeartBeatError::InvalidSequenceCounter, message.get_source_control_function());
+				}
+				else if ((message.get_uint8_at(0) != managedHeartbeat->sequenceCounter + 1) &&
+				         (250 != managedHeartbeat->sequenceCounter))
+				{
+					LOG_ERROR("[HB]: Invalid sequence counter received in heartbeat.");
+					interface->heartbeatErrorEventDispatcher.call(HeartBeatError::InvalidSequenceCounter, message.get_source_control_function());
+				}
+				else if ((250 == managedHeartbeat->sequenceCounter) &&
+				         (0 != message.get_uint8_at(0)))
+				{
+					LOG_ERROR("[HB]: Invalid sequence counter received in heartbeat.");
+					interface->heartbeatErrorEventDispatcher.call(HeartBeatError::InvalidSequenceCounter, message.get_source_control_function());
+				}
```
```suggestion
				else if (message.get_uint8_at(0) != ((managedHeartbeat->sequenceCounter + 1) % 250))
				{
					LOG_ERROR("[HB]: Invalid sequence counter received in heartbeat.");
					interface->heartbeatErrorEventDispatcher.call(HeartBeatError::InvalidSequenceCounter, message.get_source_control_function());
				}
```

GwnDaan reviewed (changes requested):
Some initial feedback, I haven't been able to test it yet but it looks promising

ad3154 commented on isobus/include/isobus/isobus/isobus_heartbeat.hpp:
```diff
@@ -0,0 +1,158 @@
+		/// repetition rate, or when the sequence counter is not valid.
+		/// The control function that generated the error is passed as an argument to the event.
+		/// @returns An event dispatcher for heartbeat errors
+		EventDispatcher<HeartBeatError, std::shared_ptr<ControlFunction>> &GetHeartbeatErrorEventDispatcher();
```
My work habits rubbing off on me... at work we use camel pascal case....

ad3154 commented:
> Some initial feedback, I haven't been able to test it yet but it looks promising

I think I addressed all your comments/suggestions in the latest diff, and I tested with the seeder example and it appeared to start sending a compliant heartbeat on request, which is great! The counter appeared to start at 251, roll over to 0, and reset once 250 was crossed as expected.

I also fixed the seeder example's manufacturer code (was 64) to be our code.

GwnDaan commented:
Sounds good, I re-opened one of them, but other than that looks good! I'm thinking it might be

GwnDaan reviewed (approved):
Awesome!
